### PR TITLE
fix(data): keep large OSM PBF extracts from freezing the app

### LINKED ISSUE #596: Bug: errors on using pbf data from geofabrik
I was trying to use the pbf data from https://download.geofabrik.de/

First I downloaded the data from a country (300 mb roughly) and then I was trying to add it to geolibre but after 5 minutes the app was unresponsive. The laptop has 32 gb of ram, if it matters. With smaller sets it was working, like 90 mb. In Qgis, I can load the data. What do you think the problem is?

I also tried to load the data using the link but it didn't work.

Thanks!

### PR DESCRIPTION
**Pull request by giswqs**
## Problem

Loading a whole-country OpenStreetMap `.osm.pbf` extract (~300 MB, e.g. a geofabrik country download) made the app unresponsive even on a 32 GB machine. Smaller extracts (~90 MB) worked. Fixes #596.

## Root cause

The OSM PBF parser (`osm-pbf.ts`) converted **every** node, way, and relation into a heavyweight GeoJSON `Feature` and held the three feature arrays in memory alongside the `Osm` index. On a country extract that is millions of JS objects, which silently OOM-kills the parse worker; the existing 5-minute timeout then fires and the UI looks frozen.

I checked whether the DuckDB-WASM Spatial path (used for other vector formats) could take over, since native DuckDB reads OSM PBF fine. It can't: the WASM-bundled GDAL build fails to open `.osm.pbf` with `IO Error: Could not parse configuration file for OSM import` (the OSM driver's `osmconf.ini` resource isn't present in the WASM build). So the fix stays in the osmix path.

## Changes

- **Emit only tagged ways and relations**, mirroring the existing tagged-only rule for nodes. Untagged ways are almost always relation-member geometry (rings/segments), not standalone features; they remain in the index for relation/polygon assembly but are no longer materialized as features. This cuts both RAM and the number of features MapLibre has to render, and matches `osmtogeojson` behavior. Real features (roads, buildings, areas, routes) carry tags and are untouched.
- **Worker progress reporting** through the classification phase, so a large extract shows live "*x* of *y* OSM entities processed" instead of a static, frozen-looking dialog.
- **Graceful too-large failure**: a parse timeout or worker OOM now surfaces an actionable "this extract is too large to parse in the browser, pre-filter it first (e.g. with osmium-tool or QGIS)" message in the Add Data dialog and in both drag-and-drop paths, instead of a silent freeze.

The remote-URL note already explains that browser loads need a CORS-enabled host (the reporter's raw geofabrik URL is CORS-blocked in the web build); that guidance is unchanged.

## Tests

- New unit tests: untagged ways are skipped (not emitted as features), and progress is reported through the parse. Existing tests unchanged.
- Verified in the running app with Playwright using a real OSM extract, in both light and dark themes: the file loads into points/lines/polygons layers, the dialog renders correctly in both themes, and the console is error-free.
- `npm run build` and `npm run test:frontend` (1228 tests) pass; pre-commit (including the npm-build hook) is clean.

## Summary by CodeRabbit

* **New Features**
  * Added real-time progress updates (processed/total) during OSM PBF import/parsing.
  * Improved user guidance for oversized OSM PBF files that can’t be parsed in-browser.
* **Bug Fixes**
  * Improved handling so untagged OSM geometries aren’t incorrectly emitted as standalone features.
* **Tests**
  * Expanded OSM PBF fixture set and added assertions for progress reporting and untagged-geometry edge cases.

### CODE REVIEW

File: apps/geolibre-desktop/src/components/layout/DesktopShell.tsx [hunk 1]
```diff
@@ -826,7 +827,9 @@ export function DesktopShell({
                   // rest of the drop.
                   setDropMessage(null);
                   setDropError(
-                    `Could not parse ${name}: ${err instanceof Error ? err.message : String(err)}`,
+                    err instanceof OsmPbfTooLargeError
+                      ? `${name} is too large to parse in the browser. Try a smaller region or pre-filter the file first (for example with osmium-tool or QGIS).`
+                      : `Could not parse ${name}: ${err instanceof Error ? err.message : String(err)}`,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/layout/DesktopShell.tsx:832]: Done in cf8ba2a — both drop-handler too-large errors now use `t("toolbar.error.osmPbfTooLarge")` (introduced in 3b54bd0, filename prefix dropped in cf8ba2a).

File: apps/geolibre-desktop/src/lib/osm-pbf.ts [hunk 2]
```diff
@@ -145,16 +170,31 @@ export async function parseOsmPbf(bytes: Uint8Array): Promise<OsmPbfLayers> {
...
   for (const way of osm.ways) {
+    tick();
+    if (!hasTags(way.tags)) continue;
     place(osmEntityToGeoJSONFeature(osm, way) as Feature | null);
   }
   for (const relation of osm.relations) {
+    tick();
+    if (!hasTags(relation.tags)) continue;
     place(osmEntityToGeoJSONFeature(osm, relation) as Feature | null);
   }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf.ts:201]: Done in 3b54bd0 — added a tagged+untagged relation fixture (`untagged-relation.osm.pbf`) and a test asserting only the tagged relation becomes a feature.

File: apps/geolibre-desktop/src/lib/osm-pbf.ts [hunk 3]
```diff
@@ -145,16 +170,31 @@ export async function parseOsmPbf(bytes: Uint8Array): Promise<OsmPbfLayers> {
...
+  onProgress?.({ processed, total });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf.ts]: Done in 3b54bd0 (refined in cf8ba2a) — the final `onProgress` is now guarded with `total > 0 && total % PROGRESS_INTERVAL !== 0`, so it no longer double-fires when total is a multiple of the interval.

File: apps/geolibre-desktop/src/lib/osm-pbf.ts [hunk 4]
```diff
@@ -28,6 +28,21 @@ export interface OsmPbfLayers {
   };
 }
 
+/** Progress through the synchronous classification phase of a parse. */
+export interface OsmPbfProgress {
+  /** Entities classified into layers so far. */
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf.ts]: Done in ce95e1c — the `OsmPbfProgress.processed` docstring now states it counts every entity visited (including skipped untagged ones), not just emitted features.

File: apps/geolibre-desktop/src/hooks/useOsmPbfLoader.ts [hunk 5]
```diff
@@ -4,8 +4,10 @@ import {
   addOsmPbfLayers,
   loadOsmPbf,
   osmPbfBaseName,
+  OsmPbfTooLargeError,
   OSM_PBF_SIZE_WARN_BYTES,
 } from "../lib/osm-pbf-loader";
+import type { OsmPbfProgress } from "../lib/osm-pbf-loader";
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/hooks/useOsmPbfLoader.ts]: Done in ce95e1c — merged the `OsmPbfProgress` type import into the existing import block.

File: apps/geolibre-desktop/src/components/layout/DesktopShell.tsx [hunk 6]
```diff
@@ -826,7 +827,9 @@ export function DesktopShell({
                   // rest of the drop.
                   setDropMessage(null);
                   setDropError(
-                    `Could not parse ${name}: ${err instanceof Error ? err.message : String(err)}`,
+                    err instanceof OsmPbfTooLargeError
+                      ? `${name}: ${t("toolbar.error.osmPbfTooLarge")}`
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/layout/DesktopShell.tsx]: Done in cf8ba2a — dropped the `${name}:` prefix; the too-large branch now shows just `t("toolbar.error.osmPbfTooLarge")`, matching the dialog path.

File: apps/geolibre-desktop/src/components/layout/DesktopShell.tsx [hunk 7]
```diff
@@ -941,7 +944,9 @@ export function DesktopShell({
             // of the drop (including any co-dropped non-PBF files).
             setDropMessage(null);
             setDropError(
-              `Could not parse ${file.name}: ${err instanceof Error ? err.message : String(err)}`,
+              err instanceof OsmPbfTooLargeError
+                ? `${file.name}: ${t("toolbar.error.osmPbfTooLarge")}`
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/layout/DesktopShell.tsx]: Done in cf8ba2a — same prefix removal on the browser path. The drop paths intentionally keep the plain "Parsing …" message (no onProgress) to stay consistent with the other non-i18n drop-overlay strings; the i18n-backed progress lives in the Add Data dialog.

File: apps/geolibre-desktop/src/lib/osm-pbf-loader.ts [hunk 8]
```diff
@@ -108,15 +131,25 @@ export function loadOsmPbf(
     worker.addEventListener(
       "message",
       (event: MessageEvent<OsmPbfWorkerMessage>) => {
-        finish();
         const data = event.data;
+        // Progress messages (the only ones carrying `type`) stream in before
+        // the single terminal message; keep the worker alive and forward them.
+        if (data && "type" in data) {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf-loader.ts]: Done in cf8ba2a — the handler now identifies the terminal message by `"ok" in data` first, then treats a `type === "progress"` message as progress, so a future field on the success/failure shapes cannot be misclassified.

File: apps/geolibre-desktop/src/lib/osm-pbf-loader.ts [hunk 9]
```diff
@@ -108,15 +131,25 @@ export function loadOsmPbf(
...
         else reject(new Error(data?.error || "Could not parse the OSM PBF file."));
       },
     );
     worker.addEventListener("error", (event) => {
       finish();
-      reject(new Error(event.message || "The OSM PBF worker failed."));
+      // A worker `error` on a large parse is usually a browser out-of-memory
+      // kill, so surface the same actionable "too large" guidance as a timeout.
+      reject(
+        new OsmPbfTooLargeError(event.message || "The OSM PBF worker failed."),
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf-loader.ts]: Done in cf8ba2a — reverted the worker `error` handler to reject with a plain `Error`; the timeout remains the sole `OsmPbfTooLargeError` signal (an OOM kill fires no `error` event, so the timeout is what catches it).

File: apps/geolibre-desktop/src/components/layout/DesktopShell.tsx [hunk 10]
```diff
@@ -806,7 +807,11 @@ export function DesktopShell({
                           bytes.byteOffset,
                           bytes.byteOffset + bytes.byteLength,
                         ) as ArrayBuffer);
-                  const layers = await loadOsmPbf(buffer);
+                  const layers = await loadOsmPbf(buffer, undefined, (p) =>
+                    setDropMessage(
+                      `Parsing ${name}… ${p.processed.toLocaleString()} of ${p.total.toLocaleString()} entities`,
+                    ),
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/layout/DesktopShell.tsx]: Done in cf8ba2a — the inline drop-progress strings were removed; the drop paths use the existing plain "Parsing …" message, so no new untranslated UI strings remain.

File: apps/geolibre-desktop/src/lib/osm-pbf.ts [hunk 11]
```diff
@@ -145,16 +175,35 @@ export async function parseOsmPbf(bytes: Uint8Array): Promise<OsmPbfLayers> {
...
     place(osmEntityToGeoJSONFeature(osm, way) as Feature | null);
   }
   for (const relation of osm.relations) {
+    tick();
+    if (!hasTags(relation.tags)) continue;
     place(osmEntityToGeoJSONFeature(osm, relation) as Feature | null);
   }
+  // Ensure a final 100% update, but skip it when the last tick already fired one
+  // for processed === total (i.e. total is an exact multiple of the interval).
+  if (total === 0 || total % PROGRESS_INTERVAL !== 0) {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf.ts]: Done in cf8ba2a — the final progress guard is now `total > 0 && total % PROGRESS_INTERVAL !== 0`, so an empty extract no longer emits a `{0,0}` ping.

File: apps/geolibre-desktop/src/i18n/locales/en.json [hunk 12]
```diff
@@ -1126,6 +1127,7 @@
       "osmPbfNoFeatures": "No point, line, or polygon features were found in this OSM PBF file.",
       "couldNotLoadOsmPbf": "Could not load the OSM PBF file.",
       "osmPbfLoadFailedWithHint": "{{message}} If this is a Mapbox Vector Tile (.pbf), it cannot be loaded as an OSM extract.",
+      "osmPbfTooLarge": "This OSM extract is too large to parse in the browser and ran out of memory or time. Try a smaller region, or pre-filter the file first (for example with osmium-tool or QGIS) before loading it.",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/i18n/locales/en.json]: Done in cf8ba2a — reworded to "This OSM extract is too large to load …" so it reads correctly in the desktop webview as well as the browser.

File: apps/geolibre-desktop/src/components/layout/DesktopShell.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/layout/DesktopShell.tsx]: Done in cf8ba2a — the inline drop-progress strings were reverted to the plain "Parsing …" message, consistent with the surrounding (non-i18n) drop-overlay messages; the i18n-backed progress lives in the Add Data dialog.

File: apps/geolibre-desktop/src/i18n/locales/en.json [hunk 13]
```diff
@@ -1126,6 +1127,7 @@
       "osmPbfNoFeatures": "No point, line, or polygon features were found in this OSM PBF file.",
       "couldNotLoadOsmPbf": "Could not load the OSM PBF file.",
       "osmPbfLoadFailedWithHint": "{{message}} If this is a Mapbox Vector Tile (.pbf), it cannot be loaded as an OSM extract.",
+      "osmPbfTooLarge": "This OSM extract is too large to load and ran out of memory or time. Try a smaller region, or pre-filter the file first (for example with osmium-tool or QGIS) before loading it.",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/i18n/locales/en.json:1130]: Intentional per CLAUDE.md / docs/i18n.md: `en.json` is the source of truth and missing keys fall back to English at runtime (no build or type break). Adding English text to the other locale files would masquerade as translated and discourage real translation. Leaving the two keys for a translation follow-up.

File: apps/geolibre-desktop/src/lib/osm-pbf-loader.ts [hunk 14]
```diff
@@ -93,7 +116,7 @@ export function loadOsmPbf(
     );
     const timeout = setTimeout(() => {
       worker.terminate();
-      reject(new Error("OSM PBF parsing timed out."));
+      reject(new OsmPbfTooLargeError("OSM PBF parsing timed out."));
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/osm-pbf-loader.ts:119]: Acknowledged — this is in tension with the earlier thread (resolved in cf8ba2a) asking not to blanket-map worker `error` events to `OsmPbfTooLargeError`. I kept the plain-`Error` mapping as the conservative default so a genuine worker crash is not mislabeled "too large"; OOM is not reliably distinguishable from other errors cross-browser. The Safari/WebKit OOM-via-`error` case is a real edge-case tradeoff, so I am leaving this open for @giswqs to decide whether WebKit-desktop parity is worth mapping `error` to `OsmPbfTooLargeError`.
